configmerge/file_reader.go

```diff
@@ -0,0 +1,109 @@
+package configmerge
+
+import (
+	"io"
+	"os"
+
+	logV2 "github.com/bitrise-io/go-utils/v2/log"
+	"github.com/go-git/go-billy/v5/memfs"
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/storage/memory"
+)
+
+type fileReader struct {
+	logger logV2.Logger
+}
+
+func NewFileReader(logger logV2.Logger) FileReader {
+	return fileReader{
+		logger: logger,
+	}
+}
+
+func (f fileReader) ReadFileFromFileSystem(name string) ([]byte, error) {
+	file, err := os.Open(name)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err := file.Close(); err != nil {
+			f.logger.Warnf("Failed to close file: %s", err)
+		}
+	}()
+	return io.ReadAll(file)
+
+}
+
+func (f fileReader) ReadFileFromGitRepository(repository string, branch string, commit string, tag string, path string) ([]byte, error) {
+	opts := git.CloneOptions{
+		URL: repository,
+	}
+	if branch != "" {
+		opts.ReferenceName = plumbing.NewBranchReferenceName(branch)
+	}
+
+	repo, cloneErr := git.Clone(memory.NewStorage(), memfs.New(), &opts)
```

Comment:
Would it make sense to introduce a local cache for the cloned repos? That way we do not need to perform the same action in the same merge command?

Reply:
We agreed that a local file system cache should be used for a config merge process. This cache might be a folder in tmp dir and gets cleaned up, once the given bitrise.yml is resolved and a new merge for the same config would always start with a new cache.